core/src/SYCL/Kokkos_SYCL_Instance.hpp

```diff
@@ -70,18 +70,206 @@ class SYCLInternal {
+    USMObjectMem& operator=(USMObjectMem&&) = delete;
+    USMObjectMem& operator=(USMObjectMem const&) = delete;
+
+    ~USMObjectMem() {
+      assert(!m_size);
+
+      if (m_data) sycl::free(m_data, *m_q);
+    }
+
+    void* data() noexcept { return m_data; }
+    const void* data() const noexcept { return m_data; }
+
+    size_t size() const noexcept { return m_size; }
+    size_t capacity() const noexcept { return m_capacity; }
+
+    // reserve() allocates space for at least n bytes
+    // returns the new capacity
+    size_t reserve(size_t n) {
+      assert(!m_size);
+      assert(m_q);
+
+      if (m_capacity < n) {
+        // First free what we have (in case malloc can reuse it)
+        sycl::free(m_data, *m_q);
+
+        m_data = sycl::malloc(n, *m_q, kind);
+        if (!m_data) {
+          m_capacity = 0;
+          throw std::bad_alloc();
+        }
+
+        m_capacity = n;
+      }
+
+      return m_capacity;
+    }
+
+   private:
+    // This will memcpy an object T into memory held by this object
+    // returns: a T* to that object
+    //
+    // Note:  it is UB to dereference this pointer with an object that is
+    // not an implicit-lifetime nor trivially-copyable type, but presumably much
+    // faster because we can use USM device memory
+    template <typename T>
+    std::unique_ptr<T, Deleter> memcpy_from(const T& t) {
+      reserve(sizeof(T));
+      sycl::event memcopied = m_q->memcpy(m_data, std::addressof(t), sizeof(T));
+      memcopied.wait();
+
+      std::unique_ptr<T, Deleter> ptr(reinterpret_cast<T*>(m_data),
+                                      Deleter(this));
+      m_size = sizeof(T);
+      return ptr;
+    }
+
+    // This will copy-constuct an object T into memory held by this object
+    // returns: a unique_ptr<T, destruct_delete> that will call the
+    // destructor on the type when it goes out of scope.
+    //
+    // Note:  This will not work with USM device memory
+    template <typename T>
+    std::unique_ptr<T, Deleter> copy_construct_from(const T& t) {
+      static_assert(kind != sycl::usm::alloc::device,
+                    "Cannot copy construct into USM device memory");
+
+      reserve(sizeof(T));
+
+      std::unique_ptr<T, Deleter> ptr(new (m_data) T(t), Deleter(this));
+      m_size = sizeof(T);
+      return ptr;
+    }
+
+   public:
+    // Performs either memcpy (for USM device memory) and returns a T*
+    // (but is technically UB when dereferenced on an object that is not
+    // an implicit-lifetime nor trivially-copyable type
+    //
+    // or
+    //
+    // performs copy construction (for other USM memory types) and returns a
+    // unique_ptr<T, ...>
+    template <typename T>
+    std::unique_ptr<T, Deleter> copy_from(const T& t) {
+      if constexpr (sycl::usm::alloc::device == kind)
+        return memcpy_from(t);
+      else
+        return copy_construct_from(t);
+    }
+
+   private:
+    template <typename T>
+    T& memcpy_to(T& t) {
```

Comment:
Why do you hav both a return value and an output parameter?

Reply:
In debugging it was useful (I could assert the memcpy_to did the right thing in one line).  Would you rather I just make it `void`?

Reply:
Added comment